Partials guide (#5128)
withastro/docs: src/content/docs/en/guides/partials.mdx

---
title: Partials
description: Learn how to create partial pages for use with dynamic front-end libraries.
i18nReady: true
---

import Badge from '~/components/Badge.astro';
import Since from '~/components/Since.astro'

<Since v="3.4.0" /> <Badge>Experimental</Badge>

> Astro supports pages that are rendered as *partials*. With partials you can load part of a page dynamically in the client without changing the URL with the aide of client-side code. When Astro renders a partial page it omits the `<doctype>` declaration and any `<head>` content, such as scoped styles and scripts.

Reviewer: Astro supports pages that are rendered as partials. With partials you can load part of a page dynamically on the client-side without changing the URL, with the help of client-side code. When Astro renders a partial page, it omits the <doctype> declaration and any <head> content, such as scoped styles and scripts.

Partials are created as regular pages in the `src/pages` directory, but are configured to be partials with this export:

```astro
---
export const partial = true;
---

<li>I'm a partial!</li>
```

The `export const partial` must be identifiable statically. It can have the value of:

- The boolean __`true`__.
- An environment variable using import.meta.env such as `import.meta.env.USE_PARTIALS`.

## Constraints

> Partials are often used in conjuction with frontend libraries such as [htmx](https://htmx.org/) and [Unpoly](https://unpoly.com/). You can also use them if you are comfortable writing low-level front-end JavaScript. For this reason they are an advanced feature.

Reviewer: ```suggestion
Partials are often used in conjunction with frontend libraries such as [htmx](https://htmx.org/) and [Unpoly](https://unpoly.com/). You can also use them if you are comfortable writing low-level front-end JavaScript. For this reason they are an advanced feature.
```

> Additionally partials should not be used if the component contains scoped styles or scripts. Partials strip out these things from the HTML output. If you need scoped styles then it is better to use regular, non-partial pages, along with a frontend library that knows how to merge the contents into the head.

Reviewer: ```suggestion
Additionally, partials should not be used if the component contains scoped styles or scripts, as partials strip out these elements from the HTML output. If you need scoped styles, it is better to use regular, non-partial pages along with a frontend library that knows how to merge the contents into the head.
```
